4512626: Non-editable JTextArea provides no visual indication of keyboard focus

azuev-java opened the pull request:
Set the text caret to be visible but not blinking on the non-editable text area. Fix the regression test that becames unstable after the change.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-4512626](https://bugs.openjdk.org/browse/JDK-4512626): Non-editable JTextArea provides no visual indication of keyboard focus
 * [JDK-8194048](https://bugs.openjdk.org/browse/JDK-8194048): Regression automated test '/open/test/jdk/javax/swing/text/DefaultCaret/HidingSelection/HidingSelectionTest.java' fails
 * [JDK-8213562](https://bugs.openjdk.org/browse/JDK-8213562): Test javax/swing/text/DefaultCaret/HidingSelection/MultiSelectionTest.java fails


### Reviewers
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**) ⚠️ Review applies to [9463a0bb](https://git.openjdk.org/jdk20/pull/21/files/9463a0bbe55871fdd30d17bc4ff1606815f167ae)
 * [Alexander Zvegintsev](https://openjdk.org/census#azvegint) (@azvegint - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk20 pull/21/head:pull/21` \
`$ git checkout pull/21`

Update a local copy of the PR: \
`$ git checkout pull/21` \
`$ git pull https://git.openjdk.org/jdk20 pull/21/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21`

View PR using the GUI difftool: \
`$ git pr show -t 21`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk20/pull/21.diff">https://git.openjdk.org/jdk20/pull/21.diff</a>

</details>

azuev-java committed:
4512626: Non-editable JTextArea provides no visual indication of keyboard focus

Set the text caret to be visible but not blinking on the non-editable text area.
Fix the regression test that becames unstable after the change.

azuev-java commented:
Continuation of https://github.com/openjdk/jdk/pull/11408 to be integrated into JDK20 workspace.

prsadhuk commented on test/jdk/javax/swing/text/DefaultCaret/HidingSelection/HidingSelectionTest.java:
```diff
@@ -119,35 +117,12 @@ public static void main(String[] args) throws Exception {
         robot.waitForIdle();
         robot.delay(200);
-        if (biEqual(robot.createScreenCapture(bounds), nosel)) {
+        if (!field2.getCaret().isSelectionVisible()) {
```
Does it need similar change in test/jdk/javax/swing/text/DefaultCaret/HidingSelection/MultiSelectionTest.java
as well?
Since the logic is changed, I will still request to make this test stable on windows (now that we are not using screen capture) and then deproblemlist

azuev-java replied:
> Does it need similar change in test/jdk/javax/swing/text/DefaultCaret/HidingSelection/MultiSelectionTest.java as well? Since the logic is changed, I will still request to make this test stable on windows (now that we are not using screen capture) and then deproblemlist

I haven't tested the MultiSelectionTest but since it is problemlisted on Linux similar change might make it more stable. As of this test - since you are the owner of JDK-8194048 i thought that you can review my changes and deproblemlist the test if it is stable on the rest of the platforms and if changes make sense. I definitely do not want to steal your thunder. As of MultiSelectionTest i can take care of it either here or in the original bug that is noted in ProblemList.txt since my fix does not affect it.

prsadhuk replied:
I guess anyway one is supposed to do a CI job for the fix and test the regression test on all platforms, so if the modified test is problemlisted it will not be tested in CI job, so it's better to deproblemlist and test...
You can do /issue add of that problemlisted bug (it just happens to be in my name for the time being, nothing stopping you assigning to you since you did all the hard work)

azuev-java replied:
I have fixed the second test and removed both of them from the problem list.

aivanov-jdk reviewed (approved):
It would be better to verify the updated test passes on Windows. If so, remove it from the problem list.

azuev-java commented:
/issue add 8194048

azuev-java commented:
/issue add 8213562

azuev-java committed:
Fixed MultiSelectionText so it is now stable on Linux
Removed both fixed tests from ProblemList

azuev-java commented:
> It would be better to verify the updated test passes on Windows. If so, remove it from the problem list.

Done and done.

azvegint commented on src/java.desktop/share/classes/javax/swing/text/DefaultCaret.java:
```diff
@@ -1032,10 +1043,14 @@ public void setVisible(boolean e) {
+                if (flasher == null) {
+                    flasher = new Timer(rate, handler);
+                }
+                flasher.setDelay(rate);
```
Coming back to negative rate, looks like this changes the previous behavior:

Before the fix it did throw IAE in all cases:

https://github.com/openjdk/jdk20/blob/ca39eb906692568347e7f264520593188f9276cf/src/java.desktop/share/classes/javax/swing/Timer.java#L397-L406

After the fix IAE is thrown only if component is editable<br>
In case of non-editable and may throw it way later after `setEditable(true)` call and focus gain.

azuev-java replied:
Fixed.

azvegint commented on src/java.desktop/share/classes/javax/swing/text/DefaultCaret.java:
```diff
@@ -379,8 +380,18 @@ protected void moveCaret(MouseEvent e) {
+                }
+            } else {
+                if (getBlinkRate() != 0) {
+                    savedBlinkRate = getBlinkRate();
```
Look like this can override blink rate set by user:

```java
import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JTextField;
import javax.swing.SwingUtilities;
import java.awt.BorderLayout;
import java.awt.Component;
import java.awt.Robot;
import java.lang.reflect.InvocationTargetException;

public class CaretThing {

    static JFrame frame;
    static JTextField textField;
    static Robot robot;
    static JButton button;

    static void printBlinkRate() {
        System.err.println("Caret blink rate: " + textField.getCaret().getBlinkRate());
    }

    static void requestFocus(Component component) throws InterruptedException, InvocationTargetException {
        robot.waitForIdle();
        robot.delay(300);

        SwingUtilities.invokeAndWait(() -> {
            System.err.println("Requesting focus on " + component);
            printBlinkRate();
            component.requestFocus();
            printBlinkRate();
        });
    }

    static void changeEditable() throws InterruptedException, InvocationTargetException {
        robot.waitForIdle();
        robot.delay(300);
        SwingUtilities.invokeAndWait(()-> {
            System.err.println("Changing editable");
            printBlinkRate();
            textField.setEditable(!textField.isEditable());
            printBlinkRate();
        });
    }

    public static void main(String[] args) throws Exception {
        robot = new Robot();
        robot.setAutoWaitForIdle(true);
        robot.setAutoDelay(50);

        SwingUtilities.invokeAndWait(() -> {
            frame = new JFrame("Hey");
            frame.setLocationRelativeTo(null);
            textField = new JTextField("Some long field value");

            textField.setEditable(false);
            frame.setLayout(new BorderLayout());
            frame.add(textField, BorderLayout.NORTH);

            button = new JButton("Button");
            button.addActionListener((e)-> printBlinkRate());

            frame.add(button, BorderLayout.CENTER);
            frame.pack();
            frame.setVisible(true);

            button.requestFocus();
            textField.getCaret().setBlinkRate(2500);
        });

        requestFocus(textField);

        requestFocus(button); // comment this to get 0 blink rate on editable field

        changeEditable();

        requestFocus(textField);

        SwingUtilities.invokeAndWait(CaretThing::printBlinkRate);
    }
}
```
Running the sample above at least two possible misbehaviors(tested on Linux):
* you will receive blink rate `500` instead of user set `2500`
* but if you comment `requestFocus(button);` line then blink rate will be `0` (it shows non-blinking cursor)

<hr>

Other than that I am worried about `getBlinkRate()` does not return the same value as passed to`setBlinkRate()` (e.g. case when component is not editable, possible JCK issue?)

azuev-java replied:
I updated the code now the behaviour is correct (as far as i tested) and the blink rate is reported as if it is set to the correct value even on non-editable text element.

azvegint replied:
I am still having issue with non-blinking cursor on editable field with the latest version of the fix.

Fails for me with test provided above and commented `requestFocus(button);` line.

azvegint replied:
The latest version does not work good for me,
e.g. with the test above and `textField.getCaret().setBlinkRate(250)`(not `2500`):

1. with `requestFocus(button);`:
 caret blinks several times and stops
 on mouse move, focus gained on text field the caret blinks several times and stops again.
2. without `requestFocus(button);`:
 it doesn't blink until focus out / focus in on the text component
 after this blinks several times and stops.

azuev-java replied:
That's weird, does not happen on any of my systems. Which OS are you testing on?

azvegint replied:
Ubuntu 22.04 and it reproduces always.

azuev-java replied:
Ok, i tried on Ubuntu 22 on x11 desktop and i see this behavior. However - i checked out master and i also can see the same glitch so this is not related to my fix.

azvegint replied:
Can confirm, ~~filed [JDK-8299050](https://bugs.openjdk.org/browse/JDK-8299050) and [JDK-8299051](https://bugs.openjdk.org/browse/JDK-8299051) to track them~~.

P.S. Just saw that you already filed them as [JDK-8299047](https://bugs.openjdk.org/browse/JDK-8299047) and [JDK-8299048](https://bugs.openjdk.org/browse/JDK-8299048)

azuev-java committed:
Handle different cases where refresh rate can be set when component is in different states causing some confused result.
Also added code that in case of non-editable text component reports the saved blink rate to avoid any problems with JCK testing.

azvegint commented on src/java.desktop/share/classes/javax/swing/text/DefaultCaret.java:
```diff
@@ -366,6 +366,8 @@ protected void moveCaret(MouseEvent e) {
     }
 
+    private int savedBlinkRate = 0;
+    private boolean blinkRateSaved = false;
```
Naming of `savedBlinkRate` and `blinkRateSaved` looks almost the same and confusing for me.
Probably `isBlinkRateSaved` would be better to highlight boolean variable.

azuev-java replied:
Fixed.

azuev-java committed:
Fixing situation where we can be stuck with non-blinking cursor after editable state switched on a focused text field

azuev-java committed:
Renamed blinkRateSaved to isBlinkRateSaved

azuev-java commented:
Ok, i fixed the permanent blinking caret glitch. Now it is non-blinking until the focus is lost and regained. Which is an improvement, without this fix the caret would be invisible until the focus lost.

azuev-java commented:
/integrate
